Replace concrete KafkaConsumers with Suppliers

meenie216 opened the pull request:
KafkaConsumers are not thread safe - using the same KafkaConsumer in both the `KafkaTopicPartitionSource` and `ReadKafkaTopicPartitionRange` would result in exceptions being thrown by the consumer as it detected being used on different threads.

This change replaces the KafkaConsumers with Suppliers to step around this issue.

meenie216 committed:
Replace concrete KafkaConsumers with Suppliers

KafkaConsumers are not thread safe, and the verious operations using the
consumer could be executed on either the main thread, or grpc executor.

MarkOSIndustries commented on example/src/main/java/com/markosindustries/distroboy/example/ExampleKafkaConsumer.java:
```diff
@@ -51,21 +51,22 @@ static void runKafkaExample(Cluster cluster) {
             DistributedOpSequence.readFrom(
                     new KafkaTopicPartitionsSource(
-                        kafkaConsumer, List.of("distroboy.example.topic")))
+                        () -> new KafkaConsumer<byte[], byte[]>(kafkaConfig),
```
I like this because it improves the communication to calling code that we need a new KafkaConsumer that we "own".
I was thinking about the possibility for people to provide memoized suppliers (like Guava has) and thwart our attempts at avoiding sharing KafkaConsumers between threads. I think maybe a better idea is to simply make the Consumer configuration the constructor argument, and we'll new up the KafkaConsumer at the point where we need it.

meenie216 replied:
Nice - makes sense to get in the way of memoized suppliers!

MarkOSIndustries commented on kafka/src/main/java/com/markosindustries/distroboy/kafka/ReadKafkaTopicPartitionRange.java:
```diff
@@ -23,16 +24,16 @@
       KafkaOffsetSpec startOffsetInclusiveSpec,
       KafkaOffsetSpec endOffsetExclusiveSpec) {
-    this.kafkaConsumer = kafkaConsumer;
+    this.kafkaConsumer = kafkaConsumerSupplier.get();
```
I think we should defer `new`ing up the `KafkaConsumer` to the `KafkaTopicPartitionsIterator` constructor, rather than `new` it up here.
That way `KafkaTopicPartitionsIterator` also becomes safer to use in isolation.
Since we would then own the `KafkaConsumer` and its lifecycle, we can then also make `KafkaTopicPartitionsIterator` implement `AutoCloseable` and close the `KafkaConsumer`. It would then get closed cleanly when the `ReadTopicPartitionRange` iteration has completed, rather than being the responsibility of the calling code to decide when to close it.

meenie216 replied:
Agreed - I pondered this at the time!

Made `KafkaTopicPartitionsIterator` implement the `IteratorWithResources`

MarkOSIndustries reviewed (changes requested):
Thanks for doing this Meenie, I've made some suggestions inline....

meenie216 committed:
Pass KafkaConfig around instead of suppliers of KafkaConsumers

Widening the road to the pit of success - Instantiate Consumers where
they are needed, closing them when they're done with.

MarkOSIndustries reviewed (approved):
LGTM, thanks!
